### PR TITLE
CI: Cache config.cache across runs to speed up build

### PR DESCRIPTION
**Pull request by hugovk**
Run `./configure` with the `--config-cache` to create a cache file on Ubuntu and macOS jobs, and use https://github.com/actions/cache to re-use it between runs:

```yml
uses: actions/cache@v3
with:
  path: config.cache
  key: ${{ github.job }}-${{ hashFiles('configure') }}-${{ hashFiles('configure.ac') }}
```

When the key changes, the action will upload a fresh cache file.

So we have a unique cache file per job ID ([`github.job`](https://docs.github.com/en/actions/learn-github-actions/contexts), for example: `build_macos`, `build_ubuntu_ssltests`), and we'll also get a fresh cache file whenever the `configure` or `configure.ac` files change.

Seconds to download cache + configure:

Job | Without cache | With cache | Times faster | Seconds saved
-- | -- | -- | -- | --
check_generated_files | 22 | 4 | 5.5 | 18
build_macos | 96 | 17 | 5.6 | 79
build_ubuntu | 23 | 4 | 5.8 | 19
build_ubuntu_ssltests (1.1.1t) | 24 | 4 | 6.0 | 20
build_ubuntu_ssltests (3.0.8) | 25 | 8 | 3.1 | 17
build_ubuntu_ssltests (3.0.8) | 25 | 7 | 3.6 | 18
test_hypothesis | 24 | 4 | 6.0 | 20
build_asan | 29 | 5 | 5.8 | 24
  |   |   |   |  
Total: | 268 | 53 | 5.1 | 215
  | [link](https://github.com/python/cpython/actions/runs/5058962194/attempts/1) | [link](https://github.com/python/cpython/actions/runs/5058962194)

### CODE REVIEW

**Comment by itamaro** [comment]
where is the cache file stored? is it using per-repo or per-org storage?
wondering about that, thinking how it will behave in forks.

I'm not familiar with config.cache - are the configure and configure.ac files the only relevant inputs?
what's the impact on total CI runtime when there's a cache hit? cache miss?

I also tried thinking about cross-branch and cross-PR interactions, but since the key uses the hash of the input files then if the hashes are identical across branches and PRs then it should be ok to reuse the cached file.

one concern is implicit inputs that are not captured by the current key:
- the "job name" is a proxy to flags that affect the result, but it is possible that flags change and the job name stays the same
- the versions of the github actions used may change and implicitly affect the result, but afaik that is not captured in the cache key

**Comment by hugovk** [comment]
> where is the cache file stored? is it using per-repo or per-org storage? wondering about that, thinking how it will behave in forks.

https://github.com/actions/cache does it per-org, so it won't be shared with forks. Forks will have their own copies.
 
> I'm not familiar with config.cache - are the configure and configure.ac files the only relevant inputs?

Let's ask @erlend-aasland :)

> what's the impact on total CI runtime when there's a cache hit? cache miss?

I ran this PR twice: once with no cache, and second time with cache present.

* See the "Post Restore config.cache" steps at the end for uploading a cache in the [first run](https://github.com/python/cpython/actions/runs/5058962194/attempts/1): they take either 0 or 1 seconds.

* See the "Restore config.cache" steps at the start for restoring a cache in the [second run](https://github.com/python/cpython/actions/runs/5058962194): they take either 0 or 1 seconds.

> I also tried thinking about cross-branch and cross-PR interactions, but since the key uses the hash of the input files then if the hashes are identical across branches and PRs then it should be ok to reuse the cached file.
> 
> one concern is implicit inputs that are not captured by the current key:
> 
> * the "job name" is a proxy to flags that affect the result, but it is possible that flags change and the job name stays the same

Could you give an example?

> * the versions of the github actions used may change and implicitly affect the result, but afaik that is not captured in the cache key

I _think_ the action will invalidate the cache itself if they change something, but I couldn't find anything to confirm. I also don't think the action's version is exposed, anyway.

**Comment by itamaro** [comment]
@arhadthedev provided examples :) 

the result can also be affected by the exact OS version (I don't know how granular `runner.os` is), which toolchain is used (gcc/clang), toolchain versions, and host architecture. any changes in any of these should invalidate the cache.

regarding the implementation - hashFiles can operate on multiple files, so use `hashFiles(file1, file2)` instead.
also for a given commit, the hash of these files is constant - would it be possible to extract the hash computation to perform it once and share between jobs?

**Comment by itamaro** [comment]
LG!

> I added the workflow file to hashFiles, so it will invalidate and make a new cache when things defined in the file change, for example the env vars and CLI flags.

this will not be over-conservative, invalidating for any change in the workflow file, even if it doesn't affect the result (which is probably most changes). probably a fair trade off (preferring correctness over CI latency).

> Possibly but I'm not sure if it's worth the complexity: we can just have hashFiles computed each time.

agreed. I'd suggest at least refactoring the list of files to be a shared constant.

**Review by erlend-aasland — APPROVED** [review]
LGTM. With the workflow file in the cache key, we're pretty safe regarding any environment change.

**Comment by erlend-aasland** [comment]
I guess we want this CI speedup in the other branches as well.

**Comment by hugovk** [comment]
> I'd suggest at least refactoring the list of files to be a shared constant.

Do you know how to do this? I tried a few things, but didn't get it working.

**Comment by itamaro** [comment]
> Do you know how to do this? I tried a few things, but didn't get it working.

I think workflow-level environment variable can work for this

https://docs.github.com/en/actions/learn-github-actions/variables

**Comment by hugovk** [comment]
I tried that like:

```yml
env:
  CONFIG_CACHE_HASH_FILES: ${{ hashFiles('configure', 'configure.ac', '.github/workflows/build.yml') }}
```
and:
```yml
          key: ${{ github.job }}-${{ runner.os }}-${{ env.CONFIG_CACHE_HASH_FILES }}
```

But it said "The workflow is not valid ", "Unrecognized function: 'hashFiles'"

I also tried:

```yml
env:
  CONFIG_CACHE_HASH_FILES: ('configure', 'configure.ac', '.github/workflows/build.yml')
```
and:
```yml
          key: ${{ github.job }}-${{ runner.os }}-${{ hashFiles(env.CONFIG_CACHE_HASH_FILES) }}
```

But it couldn't read the env var:
```
Cache not found for input keys: check_generated_files-Linux-
```

Let me try again with `format()`:

https://github.com/orgs/community/discussions/25761#discussioncomment-4626908

**Comment by hugovk** [comment]
Hmm, no luck, because it formats into strings, and we need a sort of tuple of strings.

We _might_ be able to do something like:

```yml
env:
  CONFIG_CACHE_HASH_FILE1: 'configure'
  CONFIG_CACHE_HASH_FILE2: 'configure.ac'
  CONFIG_CACHE_HASH_FILE3: '.github/workflows/build.yml'
```
And then format it something like this:

```yml
          key: ${{ github.job }}-${{ runner.os }}-${{ hashFiles(format('{0}', env.CONFIG_CACHE_HASH_FILE1))-${{ hashFiles(format('{0}', env.CONFIG_CACHE_HASH_FILE2))-${{ format('{0}', env.CONFIG_CACHE_HASH_FILE3)) }}
```

But it feels like a backward refactor...

**Comment by itamaro** [comment]
> But it feels like a backward refactor...

yeah, I agree, this doesn't improve anything. let's leave it as explicit lists of files, I might give it another try separately after this merges.

**Comment by erlend-aasland** [comment]
Backport through all security branches?

**Comment by erlend-aasland** [comment]
> Actually, maybe we should skip the security branches?
> 
> They've already diverged quite a bit, and don't get built (or backported) as often, so there's less benefit.

Yeah, makes sense.

**Comment by itamaro** [comment]
> I might give it another try separately after this merges.

I think it worked out nicely in gh-105008
